### ISSUE TITLE
Using hostDirectives inside Directives easily leads to error NG0309

### ISSUE THREAD

**Issue by TomTomB** [angular/angular#48128]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

By using the shiny new `hostDirectives` API, I've found myself easily running into error NG0309.
```
NG0309: Directive MyDirective matches multiple times on the same element. Directives can only match an element once.
```
In my case I am using a `DestroyDirective` (for handling rxjs subscriptions).
Sample repository: https://github.com/TomTomB/multiple-same-host-directive-issue

In it is a component with a template, that uses 2 directives on the same element (`app-a` and `app-b`)
```html
<div app-a app-b></div>
```

Both directives have the following logic:
```ts
@Directive({
  selector: '[app-a]',
  standalone: true,
  hostDirectives: [DestroyDirective],
})
export class ADirective {
  private readonly _destroy$ = inject(DestroyDirective).destroy$;
}
```

This results in the error mentioned above. 
The error is, of cause, not only specific to this exact use case. I fact, it's a pretty big road block when thinking the new "composition way" of reusing code. 
Also, the error is really not that helpful for explaining, why this code is not working.
This could especially be true when using 3rd party packages that use hostDirectives on their own.   

### Proposed solution

Using `hostDirecitves` inside a directive should never result in a NG0309 error.

### Alternatives considered

Removing `hostDirecitves` from the Directive decorator - but that would be a big bummer imho.

**Comment by dylhunn** [comment]
cc @crisbeto -- just for visibility

**Comment by mdudek** [comment]
I'm facing same problem, this issue is a show stopper for me at this moment.

**Comment by crisbeto** [comment]
Currently this is working as expected. Initially we were planning on de-duplicating the directives automatically, but the problem was that it was tricky to figure out when exactly the duplicated directive would have to run.

**Comment by alxhub** [comment]
I consider tricks like `DestroyDirective` to be an anti-pattern in the first place. The purpose of `hostDirectives` is not to provide some kind of injection mechanism for lifecycle hooks, but for one directive to compose behaviors from another. In that sense, it's perfectly valid to disallow multiple directives attempting to apply the same host directive.

**Comment by TomTomB** [comment]
Hey @alxhub,
I think just focusing on my example `DestroyDirective` is a litte short-sighted, since this issue applies to all directives. 
Let's use the angular/components library as an other example. They are currently using class mixins for things like a `disabled` boolean. 
So with `hostDirectives` I would assume that we should now be able to create a `CanDisableDirective`. Applying this to my issue repo would yield the same result - NG0309. So composition is a thing now, but not really.

**Comment by Harpush** [comment]
@TomTomB I think the problem is the definition of host directives. There are two options:
1. Host directives are meant to avoid adding a directive on every component usage but also retain UI code reusability.
2. Host directives are meant for reusable code without real correlation to the original directive context.

So option 1 matches all UI related directives. Tooltip makes sense only once and a directive that adds a border makes sense only once. Adding it twice and supporting it will result in odd behavior.

Option 2 is more like code composition without real UI related logic. In that case this error is really annoying. Under this falls destroy directive and disabled directive and many more non UI oriented logic.

To further explain it let's take you angular components disabled example. In this case there is no UI logic and it falls under option 2. It makes sense to reuse it without the error - but is it really a directive? Can it even be used as a directive which isn't in a component's host directives?

To summarize - it seems host directives were built with option 1 in mind. For option 2 (which I agree is missing as mixins aren't easy or fun) maybe another solution is needed.

**Comment by TomTomB** [comment]
@Harpush I agree with you. Option 2 was simply misunderstand and pushed through media as the "Angular now does composition" headline. 

Anyways, in my opinion the dev should be able to decide, if he wants a directive to be able to be used multiple times on the same element. 
In the following example for the `CanDisable` directive, I've added the option `multi: true` (don't quote me on that naming) to the `@Directive` decorator. `false` should of cause be the the default so we don't change Angular's default behavior.

```ts
@Directive({ standalone: true, multi: true })
export class CanDisable {
  @Input()
  get disabled(): boolean {
    return this._disabled;
  }
  set disabled(value: BooleanInput) {
    this._disabled = coerceBooleanProperty(value);
  }
  private _disabled = false;
}
```
An other option would be to not use directives at all, but that would mean a whole new feature which is probably a lot of work. Especially if we count in that host directives already implemented most of it.
